### ISSUE TITLE
ContinuousIntegrationBuild=true produces nondeterministic source paths by default

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn#55860]
I think this is an SDK bug. The workaround is either to add a specific SourceLink provider package reference (which may be redundant and may even be problematic or impossible) or to add a `<SourceRoot Include="$(SolutionDir)" />` item or similar. The actual directory does not matter so long as it contains all source files not already covered by a SourceRoot item.

Without the workaround, you can see absolute paths leaking from the CI image, and NuGet Package Explorer also flags the resulting assembly as failing the determinism check:
![image](https://user-images.githubusercontent.com/8040367/129943723-e947078a-d2de-4ee5-b247-c6e5c714e648.png)

Using the workaround, determinism is achieved:
![image](https://user-images.githubusercontent.com/8040367/129943798-945f6165-f181-479a-b0a2-af3e9bd7c85d.png)

If no specific SourceLink provider package is referenced in a project, the SDK should add a SourceRoot directory by default containing all source files. This should work the same way whether or not the project is even contained in a source-control repository of any kind. Alternatively, source files outside the solution directory should not be mapped and should cause build warnings when ContinuousIntegrationBuild=true.

**Comment by marcpopMSFT** [comment]
I think that ContinuousIntegrationBuild is used in Microsoft.Managed.Core.targets which comes from Roslyn.

**Comment by jaredpar** [comment]
@tmat is this the right place for the issue?

**Comment by tmat** [comment]
I'm not sure what the ask is. Setting `ContinuousIntegrationBuild` does not imply the resulting binaries will be deterministic. `ContinuousIntegrationBuild` only specifies that build is running on CI allowing other targets to adjust their behavior accordingly.

**Comment by tmat** [comment]
> the SDK should add a SourceRoot directory by default containing all source files

The SDK doesn't know what that directory is.

**Comment by clairernovotny** [comment]
Where does that come from? The issue is that without using any of the `Sourcelink.*` NuGet packages, and with `EmbedAllSources` set to `true`, the paths are still non-deterministic without the SourceRoot's set. The ask is that the current directory be set as a source root by default so the paths will be deterministic. When they're embedded, things like submodules and paths to a git repo are irrelevant.

**Comment by tmat** [comment]
The project can set `<SourceRoot Include="$(RepoRoot)" />` to make them deterministic in that case. The SDK does not have a concept of repository root without Source Link. Source Link determines that based on location of `.git` directory.

BTW, the paths are not irrelevant even when the source files are embedded in the PDB. The paths can still be used by the program via `CallerFileNameAttribute`.

**Comment by clairernovotny** [comment]
@tmat should the SDK have that concept though? Seems like determinism as a concept does not require SourceLink. And if the paths are different between builds, then it's not deterministic since the paths are local machine paths. I do think this should be an SDK concept.

**Comment by tmat** [comment]
>  Seems like determinism as a concept does not require SourceLink.

Indeed, determinism doesn't require Source Link. Setting `<SourceRoot Include="RepoRoot" />` manually in Directory.Build.props is the way to make paths deterministic in the repo. The SDK could add a concept of "repository root". But the customer would need to set it anyways in some way. How else would the SDK determine it? Perhaps the directory containing `global.json` could be considered root? If that works then SourceRoot could be initialized based on that by default.

**Comment by clairernovotny** [comment]
I guess what I'm looking for is a "secure by default" "it just works" mechanism.

**Comment by jmarolf** [comment]
It sounds like this should be moved to https://github.com/dotnet/sdk?

**Comment by jnm2** [comment]
> > 
> 
> The SDK doesn't know what that directory is. The directory containing the solution does not need to contain all the source files.
> Consider, e.g. repository layout `repo\src` containing all sources and `repo\solutions` containing solution files.

That's why I said the SDK could either:

- Pick the nearest common ancestor of all source files that aren't already accounted for by other SourceRoot items
- Or, default to the solution directory if there are any source files that aren't already accounted for by other SourceRoot items, and then add build warnings if there were source files outside the solution directory which instruct you to add SourceLink or specify one or more SourceRoot items manually.

**Comment by jnm2** [comment]
This is going to be fun.

**Comment by jaredpar** [comment]
Wanted to ping on this issue. What are the next steps here? Unclear from the convo if we believe this is an SDK issue or something we should be mitigating in the compiler.

**Comment by ite-jankl** [comment]
The [`ContinuousIntegrationBuild` docs say](https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#continuousintegrationbuild):

> Currently, setting this property to true works only if you add either a specific [SourceLink](https://github.com/dotnet/sourcelink) provider package reference or a <SourceRoot Include="$(MyDirectory)" /> item. For more information, see [dotnet/roslyn issue 55860](https://github.com/dotnet/roslyn/issues/55860).

I tested in a net10 Microsoft.NET.Sdk.Web project with `ContinuousIntegrationBuild` `true` and `DebugType` `embedded`, and path normalization works just fine (`/_/` prefixed relative paths)

I do not reference a source link provider package nor do I set the source root explicitly, in my project or dir props.

Is that documentation outdated? Or what's the relation here, what am I missing?

**Comment by Frulfump** [comment]
> The [`ContinuousIntegrationBuild` docs say](https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#continuousintegrationbuild):
> 
> > Currently, setting this property to true works only if you add either a specific [SourceLink](https://github.com/dotnet/sourcelink) provider package reference or a <SourceRoot Include="$(MyDirectory)" /> item. For more information, see [dotnet/roslyn issue 55860](https://github.com/dotnet/roslyn/issues/55860).
> 
> I tested in a net10 Microsoft.NET.Sdk.Web project with `ContinuousIntegrationBuild` `true` and `DebugType` `embedded`, and path normalization works just fine (`/_/` prefixed relative paths)
> 
> I do not reference a source link provider package nor do I set the source root explicitly, in my project or dir props.
> 
> Is that documentation outdated? Or what's the relation here, what am I missing?

Maybe due to Source Link being included and enabled by default since .NET 8 SDK?
https://github.com/dotnet/sourcelink#using-source-link-in-net-projects

**Comment by tmat** [comment]
> Maybe due to Source Link being included and enabled by default since .NET 8 SDK? https://github.com/dotnet/sourcelink#using-source-link-in-net-projects

Yes, that is correcty. The doc is outdated. Let me remove the note and also close this issue.
